In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('Alphabets_data.csv')
df.head(10)

,letter,xbox,ybox,width,height,onpix,xbar,ybar,x2bar,y2bar,xybar,x2ybar,xy2bar,xedge,xedgey,yedge,yedgex
0,T,2,8,3,5,1,8,13,0,6,6,10,8,0,8,0,8
1,I,5,12,3,7,2,10,5,5,4,13,3,9,2,8,4,10
2,D,4,11,6,8,6,10,6,2,6,10,3,7,3,7,3,9
3,N,7,11,6,6,3,5,9,4,6,4,4,10,6,10,2,8
4,G,2,1,3,1,1,8,6,6,6,6,5,9,1,7,5,10
5,S,4,11,5,8,3,8,8,6,9,5,6,6,0,8,9,7
6,B,4,2,5,4,4,8,7,6,6,7,6,6,2,8,7,10
7,A,1,1,3,2,1,8,2,2,2,8,2,8,1,6,2,7
8,J,2,2,4,4,2,10,6,2,6,12,4,8,1,6,1,7
9,M,11,15,13,9,7,13,2,6,2,12,1,9,8,1,1,8


In [3]:
df.shape

(20000, 17)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   letter  20000 non-null  object
 1   xbox    20000 non-null  int64 
 2   ybox    20000 non-null  int64 
 3   width   20000 non-null  int64 
 4   height  20000 non-null  int64 
 5   onpix   20000 non-null  int64 
 6   xbar    20000 non-null  int64 
 7   ybar    20000 non-null  int64 
 8   x2bar   20000 non-null  int64 
 9   y2bar   20000 non-null  int64 
 10  xybar   20000 non-null  int64 
 11  x2ybar  20000 non-null  int64 
 12  xy2bar  20000 non-null  int64 
 13  xedge   20000 non-null  int64 
 14  xedgey  20000 non-null  int64 
 15  yedge   20000 non-null  int64 
 16  yedgex  20000 non-null  int64 
dtypes: int64(16), object(1)
memory usage: 2.6+ MB


In [5]:
df.describe()

,xbox,ybox,width,height,onpix,xbar,ybar,x2bar,y2bar,xybar,x2ybar,xy2bar,xedge,xedgey,yedge,yedgex
count,20000.000000,20000.000000,20000.000000,20000.00000,20000.000000,20000.000000,20000.000000,20000.000000,20000.000000,20000.000000,20000.00000,20000.000000,20000.000000,20000.000000,20000.000000,20000.00000
mean,4.023550,7.035500,5.121850,5.37245,3.505850,6.897600,7.500450,4.628600,5.178650,8.282050,6.45400,7.929000,3.046100,8.338850,3.691750,7.80120
std,1.913212,3.304555,2.014573,2.26139,2.190458,2.026035,2.325354,2.699968,2.380823,2.488475,2.63107,2.080619,2.332541,1.546722,2.567073,1.61747
min,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.000000,0.000000,0.00000
25%,3.000000,5.000000,4.000000,4.00000,2.000000,6.000000,6.000000,3.000000,4.000000,7.000000,5.00000,7.000000,1.000000,8.000000,2.000000,7.00000
50%,4.000000,7.000000,5.000000,6.00000,3.000000,7.000000,7.000000,4.000000,5.000000,8.000000,6.00000,8.000000,3.000000,8.000000,3.000000,8.00000
75%,5.000000,9.000000,6.000000,7.00000,5.000000,8.000000,9.000000,6.000000,7.000000,10.000000,8.00000,9.000000,4.000000,9.000000,5.000000,9.00000
max,15.000000,15.000000,15.000000,15.00000,15.000000,15.000000,15.000000,15.000000,15.000000,15.000000,15.00000,15.000000,15.000000,15.000000,15.000000,15.00000


In [6]:
df.isnull().sum()

letter    0
xbox      0
ybox      0
width     0
height    0
onpix     0
xbar      0
ybar      0
x2bar     0
y2bar     0
xybar     0
x2ybar    0
xy2bar    0
xedge     0
xedgey    0
yedge     0
yedgex    0
dtype: int64

In [7]:
df['letter'].nunique()

26

In [8]:
df['letter'].unique()

array(['T', 'I', 'D', 'N', 'G', 'S', 'B', 'A', 'J', 'M', 'X', 'O', 'R',
       'F', 'C', 'H', 'W', 'L', 'P', 'E', 'V', 'Y', 'Q', 'U', 'K', 'Z'],
      dtype=object)

In [9]:
X = df.drop('letter', axis=1)
y = df['letter']


In [10]:
# preprocessing
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [11]:
X_scaled

array([[-1.0576983 ,  0.29187713, -1.05327668, ..., -0.21908163,
        -1.4381527 ,  0.12291107],
       [ 0.51038497,  1.5023577 , -1.05327668, ..., -0.21908163,
         0.12008142,  1.35944092],
       [-0.01230945,  1.19973756,  0.43590966, ..., -0.8656262 ,
        -0.26947711,  0.74117599],
       ...,
       [ 1.03307939,  0.59449727,  0.43590966, ...,  2.36709667,
        -0.65903564, -2.35014863],
       [-1.0576983 , -1.22122359, -0.55688123, ...,  0.42746295,
         0.50963994,  0.12291107],
       [-0.01230945,  0.59449727,  0.43590966, ..., -0.8656262 ,
        -0.65903564,  0.12291107]])

# model implementation

In [12]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_scaled,y,
                                                   test_size=0.2,
                                                   random_state=42,
                                                   stratify=y)


In [13]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
y_train_enc = le.fit_transform(y_train)
y_test_enc = le.transform(y_test)

we applied label encoding to convert the alphabet labels into numerical form suitable for neural network training

# Model Building- base

In [20]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout

In [15]:
model = Sequential()

# Input layer
model.add(Dense(128,activation='relu', input_shape=(X_train.shape[1],)))
# hidden
model.add(Dense(64,activation='relu',))
#output
model.add(Dense(64,activation='softmax'))

In [16]:
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy'])

In [17]:
history = model.fit(
            X_train,
            y_train_enc,
            epochs = 20,
            batch_size=32,
            validation_split=0.1)

Epoch 1/20
450/450 [==============================] - 3s 3ms/step - loss: 1.6029 - accuracy: 0.5826 - val_loss: 0.8573 - val_accuracy: 0.7681
Epoch 2/20
450/450 [==============================] - 1s 2ms/step - loss: 0.7285 - accuracy: 0.7915 - val_loss: 0.6343 - val_accuracy: 0.8156
Epoch 3/20
450/450 [==============================] - 1s 2ms/step - loss: 0.5598 - accuracy: 0.8392 - val_loss: 0.5142 - val_accuracy: 0.8500
Epoch 4/20
450/450 [==============================] - 1s 2ms/step - loss: 0.4626 - accuracy: 0.8631 - val_loss: 0.4349 - val_accuracy: 0.8719
Epoch 5/20
450/450 [==============================] - 1s 2ms/step - loss: 0.3908 - accuracy: 0.8848 - val_loss: 0.3892 - val_accuracy: 0.8881
Epoch 6/20
450/450 [==============================] - 2s 3ms/step - loss: 0.3383 - accuracy: 0.8989 - val_loss: 0.3337 - val_accuracy: 0.8963
Epoch 7/20
450/450 [==============================] - 1s 3ms/step - loss: 0.2951 - accuracy: 0.9133 - val_loss: 0.3216 - val_accuracy: 0.9050
Epoch 

In [18]:
test_loss,test_acc = model.evaluate(X_test,y_test_enc)

print('Test accuracy:',test_acc)
print('Test Loss:', test_loss)

125/125 [==============================] - 0s 2ms/step - loss: 0.1586 - accuracy: 0.9450
Test accuracy: 0.9449999928474426
Test Loss: 0.1586456000804901


In [19]:
from sklearn.metrics import classification_report, confusion_matrix

y_pred_probs = model.predict(X_test)
y_pred =np.argmax(y_pred_probs,axis=1)

print(confusion_matrix(y_test_enc,y_pred))
print(classification_report(y_test_enc,y_pred))

125/125 [==============================] - 0s 1ms/step
[[158   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0]
 [  0 138   0   0   1   0   0   0   0   0   1   0   0   0   1   0   0   4
    5   0   3   0   0   0   0   0]
 [  0   0 142   0   1   1   1   0   1   0   1   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0]
 [  0   3   0 149   0   1   0   0   0   1   0   1   0   1   3   0   0   1
    0   0   0   0   0   1   0   0]
 [  0   0   1   0 150   2   0   0   0   0   0   1   0   0   0   0   0   0
    0   0   0   0   0   0   0   0]
 [  0   0   0   0   0 146   1   1   1   1   0   0   0   2   0   3   0   0
    0   0   0   0   0   0   0   0]
 [  0   2   3   0   4   0 139   0   0   0   1   0   0   0   2   0   3   0
    0   0   0   1   0   0   0   0]
 [  0   3   2   4   0   0   1 122   0   0   3   0   1   1   2   1   0   3
    0   0   2   0   0   2   0   0]
 [  1   0   0   0   0   1   0   0 139   7   0   0   0   0   0   0   0   0

from above results we can see that class 7 have the lowest recall. the model struggles that to recognize the letter H or class 7 in this dataset.

The neural network achieved an overall accuracy of 94% on the test dataset, indicating strong performance for a 26-class classification problem.
Most classes show precision and recall above 90%, demonstrating consistent classification ability across alphabets.
Some classes exhibit slightly lower recall (e.g. class 7), suggesting the model occasionally confuses visually similar letters.
Overall, the confusion matrix shows strong diagonal dominance, confirming that the model correctly classifies the majority of samples.

# Hyperparameter tuning

In [22]:
model2 = Sequential()

model2.add(Dense(256, activation='relu', input_shape=( X_train.shape[1],)))
model2.add(Dropout(0.3))
model2.add(Dense(128, activation='relu'))
model2.add(Dropout(0.3))
model2.add(Dense(26, activation='softmax'))

In [23]:
model2.compile(
            optimizer='adam',
            loss='sparse_categorical_crossentropy',
            metrics=['accuracy'])

In [24]:
history2 = model2.fit(
            X_train,
            y_train_enc,
            epochs=30,
            batch_size=32,
            validation_split=0.1)

Epoch 1/30
450/450 [==============================] - 3s 4ms/step - loss: 1.5689 - accuracy: 0.5417 - val_loss: 0.8232 - val_accuracy: 0.7713
Epoch 2/30
450/450 [==============================] - 1s 3ms/step - loss: 0.8706 - accuracy: 0.7276 - val_loss: 0.6027 - val_accuracy: 0.8300
Epoch 3/30
450/450 [==============================] - 2s 3ms/step - loss: 0.6978 - accuracy: 0.7784 - val_loss: 0.5034 - val_accuracy: 0.8550
Epoch 4/30
450/450 [==============================] - 1s 3ms/step - loss: 0.6067 - accuracy: 0.8094 - val_loss: 0.4156 - val_accuracy: 0.8819
Epoch 5/30
450/450 [==============================] - 1s 3ms/step - loss: 0.5353 - accuracy: 0.8299 - val_loss: 0.3670 - val_accuracy: 0.9000
Epoch 6/30
450/450 [==============================] - 1s 3ms/step - loss: 0.4900 - accuracy: 0.8437 - val_loss: 0.3422 - val_accuracy: 0.8988
Epoch 7/30
450/450 [==============================] - 1s 3ms/step - loss: 0.4431 - accuracy: 0.8597 - val_loss: 0.3098 - val_accuracy: 0.9087
Epoch 

# Evaluation- final model

In [25]:
test_loss2, test_acc2 = model2.evaluate(X_test, y_test_enc)
print("Improved Test Accuracy:", test_acc2)


125/125 [==============================] - 0s 2ms/step - loss: 0.1131 - accuracy: 0.9640
Improved Test Accuracy: 0.9639999866485596


In [27]:
y_pred_probs2 = model2.predict(X_test)
y_pred2 = np.argmax(y_pred_probs2, axis=1)
cm2 = confusion_matrix(y_test_enc, y_pred2)
print(cm2)


125/125 [==============================] - 0s 1ms/step
[[157   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   1   0   0   0   0   0]
 [  0 145   0   0   0   0   0   1   0   0   1   0   0   0   0   0   0   1
    4   0   1   0   0   0   0   0]
 [  0   0 146   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0]
 [  0   2   0 158   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   1   0   0]
 [  0   0   2   0 149   0   0   0   0   0   0   1   0   0   0   0   0   1
    0   0   0   0   0   0   0   1]
 [  0   0   0   0   2 147   1   0   0   1   0   0   0   1   0   1   0   0
    0   1   0   1   0   0   0   0]
 [  0   1   1   0   6   0 143   1   0   0   1   0   0   0   0   0   0   1
    0   0   0   1   0   0   0   0]
 [  0   0   0   3   0   0   1 134   0   0   5   0   0   0   0   0   0   2
    0   0   2   0   0   0   0   0]
 [  0   0   0   0   0   2   1   0 139   5   0   0   0   0   0   0   0   0

In [28]:
print(classification_report(y_test_enc, y_pred2))


              precision    recall  f1-score   support

           0       0.99      0.99      0.99       158
           1       0.92      0.95      0.93       153
           2       0.97      0.99      0.98       147
           3       0.98      0.98      0.98       161
           4       0.93      0.97      0.95       154
           5       0.94      0.95      0.94       155
           6       0.95      0.92      0.93       155
           7       0.95      0.91      0.93       147
           8       0.98      0.92      0.95       151
           9       0.95      0.95      0.95       149
          10       0.93      0.97      0.95       148
          11       0.97      0.97      0.97       152
          12       1.00      0.98      0.99       158
          13       0.97      0.96      0.97       157
          14       0.99      0.96      0.97       150
          15       0.99      0.95      0.97       161
          16       0.98      0.98      0.98       157
          17       0.93    

To improve the neural network performance, the architecture was modified by increasing the number of neurons and adding dropout layers to reduce overfitting. The number of training epochs was also increased.

The tuned model achieved a test accuracy of 96.4%, compared to 94.5% from the baseline model.

The addition of dropout reduced overfitting and improved the model’s ability to generalize to unseen data. This demonstrates the importance of hyperparameter tuning in neural network optimization.

The tuned neural network achieved an overall accuracy of 96% on the test dataset. Precision, recall, and F1-score values are consistently high across all 26 classes, indicating balanced model performance.
Compared to the baseline model, the tuned model significantly improved recall for previously difficult classes, demonstrating better generalization and reduced overfitting.
The results confirm that hyperparameter tuning and dropout regularization improved the robustness of the neural network.

# Conclusion

In this project, a Neural Network model was developed to classify handwritten alphabets using the Alphabets_data.csv dataset. The dataset was explored, preprocessed, normalized, and used to train a multi-class classification model capable of identifying 26 alphabet classes.

During data exploration, it was observed that the dataset contained only numerical feature values, which made it suitable for neural network training after normalization. The target variable consisted of categorical alphabet labels, which were encoded into numerical form for model training.

The tuned model demonstrated better generalization and reduced overfitting.
The confusion matrix and classification report showed consistent performance across all alphabet classes, with significant improvement in previously challenging classes.
This confirms the importance of hyperparameter tuning in neural network optimization.

limitations:

    Neural networks require careful tuning, longer training time, and computational resources. Model performance can also depend heavily on architecture and hyperparameter choices.